File: electron/src/core/parser.ts
/**
 * CSV parser for tracker log files.
 * Supports flexible column mapping for multiple CSV formats.
 */

import { TrackerRecord } from './models';

/** Field name mappings for flexible CSV parsing */
const FIELD_MAPPINGS: Record<string, string[]> = {
  tracker_id: [
    'tracker_id', 'id', 'tracker', 'device_id', 'unit_id', 'unique_id', 'unique id',
    'report_stationid', 'user_loggerid', 'logger_id', 'loggerid', 'station_id', 'stationid',
  ],
  time: [
    'time', 'timestamp', 'datetime', 'time_local', 'received_time',
    'time local received', 'time_local_received',
    'measurement_datetime', 'measurement_receiveddatetime',
  ],
  time_gps: ['time_gps', 'gps_time', 'gps_timestamp'],
  lat: ['lat', 'latitude', 'gps_lat'],
  lon: ['lon', 'lng', 'longitude', 'gps_lon'],
  alt: ['alt', 'altitude', 'alt_m', 'altitude_m', 'gps_alt'],
  speed: ['speed', 'speed_mps', 'speed_m_s', 'gps_sog', 'sog'],
  course: ['course', 'heading', 'course_deg', 'gps_cog', 'cog'],
  hdop: ['hdop', 'dop', 'gps_hdop'],
  satellites: ['satellites', 'sats', 'num_sats', 'gps_satellites', 'sat_count', 'numsat'],
  fix_valid: [
    'fix_valid', 'fix', 'gps_fix', 'valid', 'gps_fix_valid', 'gps fix valid',
    'gps_fixvalid', 'fixvalid',
  ],
  rssi: ['rssi', 'rssi_dbm', 'signal', 'rf_rssi'],
  time_received: [
    'time_received', 'received_time', 'rx_time', 'receive_time',
    'measurement_receiveddatetime', 'received_datetime',
  ],
  baro_alt: [
    'baro_alt', 'baro_altitude', 'baro_alt_m', 'barometric_altitude', 'barometric altitude',
    'barometer_altitude',
  ],
  baro_temp: ['baro_temp', 'baro_temperature', 'temp_c', 'barometer_temperature'],
  baro_press: ['baro_press', 'baro_pressure', 'pressure_hpa', 'barometer_pressure'],
  battery_mv: ['battery_mv', 'battery', 'voltage_mv'],
};

/** Datetime formats to try when parsing */
const DATETIME_FORMATS = [
  // ISO-like formats
  /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})(?:\.(\d+))?(?:Z|[+-]\d{2}:\d{2})?$/,
  /^(\d{4})-(\d{2})-(\d{2}) (\d{2}):(\d{2}):(\d{2})(?:\.(\d+))?$/,
  // US format
  /^(\d{1,2})\/(\d{1,2})\/(\d{4}) (\d{2}):(\d{2}):(\d{2})$/,
  // European format
  /^(\d{1,2})\/(\d{1,2})\/(\d{4}) (\d{2}):(\d{2}):(\d{2})$/,
];

function parseCSVLine(line: string): string[] {
  const result: string[] = [];
  let current = '';
  let inQuotes = false;

  for (let i = 0; i < line.length; i++) {
    const ch = line[i];
    if (ch === '"') {
      if (inQuotes && i + 1 < line.length && line[i + 1] === '"') {
        current += '"';
        i++;
      } else {
        inQuotes = !inQuotes;
      }
    } else if (ch === ',' && !inQuotes) {
      result.push(current);
      current = '';
    } else {
      current += ch;
    }
  }
  result.push(current);
  return result;
}

function parseDatetime(value: string): string | null {
  if (!value) return null;
  value = value.trim();

  // Try ISO format first (most common)
  const isoMatch = value.match(
    /^(\d{4})-(\d{2})-(\d{2})[T ](\d{2}):(\d{2}):(\d{2})(?:\.(\d+))?(?:Z|[+-]\d{2}:\d{2})?$/
  );
  if (isoMatch) {
    const [, yr, mo, dy, hr, mn, sc, ms] = isoMatch;
    const msStr = ms ? `.${ms.padEnd(3, '0').slice(0, 3)}` : '.000';
    return `${yr}-${mo}-${dy}T${hr}:${mn}:${sc}${msStr}`;
  }

  // Try US format: M/D/YYYY H:M:S
  const usMatch = value.match(/^(\d{1,2})\/(\d{1,2})\/(\d{4}) (\d{2}):(\d{2}):(\d{2})$/);
  if (usMatch) {
    const [, mo, dy, yr, hr, mn, sc] = usMatch;
    return `${yr}-${mo.padStart(2, '0')}-${dy.padStart(2, '0')}T${hr}:${mn}:${sc}.000`;
  }

  // Try European format: D/M/YYYY H:M:S
  const euMatch = value.match(/^(\d{1,2})\/(\d{1,2})\/(\d{4}) (\d{2}):(\d{2}):(\d{2})$/);
  if (euMatch) {
    const [, dy, mo, yr, hr, mn, sc] = euMatch;
    return `${yr}-${mo.padStart(2, '0')}-${dy.padStart(2, '0')}T${hr}:${mn}:${sc}.000`;
  }

  // Last resort: try Date constructor
  try {
    const d = new Date(value);
    if (!isNaN(d.getTime())) {
      return d.toISOString().replace('Z', '');
    }
  } catch {
    // ignore
  }

  return null;
}

export class CSVParser {
  private fieldIndices: Map<string, number> = new Map();

  parseCSVContent(content: string): TrackerRecord[] {
    const records: TrackerRecord[] = [];
    const lines = content.split('\n').filter((l) => l.trim());

    if (lines.length < 2) return records;

    // Parse header
    const headers = parseCSVLine(lines[0]).map((h) => h.trim().toLowerCase());
    this.buildFieldIndices(headers);

    // Parse data rows
    for (let i = 1; i < lines.length; i++) {
      try {
        const row = parseCSVLine(lines[i]);
        const record = this.parseRow(row);
        if (record) records.push(record);
      } catch {
        // Skip invalid rows
      }
    }

    return records;
  }

  private buildFieldIndices(headers: string[]): void {
    this.fieldIndices.clear();

    for (const [fieldName, variants] of Object.entries(FIELD_MAPPINGS)) {
      for (const variant of variants) {
        const idx = headers.indexOf(variant);
        if (idx !== -1) {
          this.fieldIndices.set(fieldName, idx);
          break;
        }
      }
    }
  }

  private parseRow(row: string[]): TrackerRecord | null {
    // Must have tracker_id
    if (!this.fieldIndices.has('tracker_id')) return null;

    const trackerIdIdx = this.fieldIndices.get('tracker_id')!;
    if (trackerIdIdx >= row.length) return null;

    let trackerId = row[trackerIdIdx].trim();
    if (!trackerId) return null;

    // Normalize tracker_id: remove surrounding parentheses
    trackerId = trackerId.replace(/^\((.+)\)$/, '$1');

    // Parse time
    const timeLocal = this.getDatetime(row, 'time') || new Date().toISOString();
    const timeGps = this.getDatetime(row, 'time_gps');
    const timeReceived = this.getDatetime(row, 'time_received');

    // Parse position
    const lat = this.getFloat(row, 'lat');
    const lon = this.getFloat(row, 'lon');
    const altM = this.getFloat(row, 'alt');

    // Parse speed/course
    const speedMps = this.getFloat(row, 'speed');
    const courseDeg = this.getFloat(row, 'course');

    // Parse GPS quality
    const hdop = this.getFloat(row, 'hdop');
    const satellites = this.getInt(row, 'satellites');
    const fixValid = this.getBool(row, 'fix_valid');

    // Signal strength
    const rssiDbm = this.getFloat(row, 'rssi');

    // Barometer
    const baroAltM = this.getFloat(row, 'baro_alt');
    const baroTempC = this.getFloat(row, 'baro_temp');
    const baroPressHpa = this.getFloat(row, 'baro_press');

    // Battery
    const batteryMv = this.getFloat(row, 'battery_mv');

    // Calculate latency
    let latencyMs: number | null = null;
    if (timeLocal && timeReceived) {
      const localMs = new Date(timeLocal).getTime();
      const receivedMs = new Date(timeReceived).getTime();
      if (!isNaN(localMs) && !isNaN(receivedMs)) {
        latencyMs = receivedMs - localMs;
      }
    }

    return {
      tracker_id: trackerId,
      time_local_received: timeLocal,
      time_gps: timeGps,
      time_received: timeReceived,
      lat,
      lon,
      alt_m: altM,
      speed_mps: speedMps,
      course_deg: courseDeg,
      hdop,
      satellites,
      rssi_dbm: rssiDbm,
      baro_alt_m: baroAltM,
      baro_temp_c: baroTempC,
      baro_press_hpa: baroPressHpa,
      fix_valid: fixValid,
      battery_mv: batteryMv,
      latency_ms: latencyMs,
    };
  }

  private getFloat(row: string[], field: string): number | null {
    if (!this.fieldIndices.has(field)) return null;
    const idx = this.fieldIndices.get(field)!;
    if (idx >= row.length) return null;
    const val = row[idx].trim();
    if (!val) return null;
    const num = parseFloat(val);
    return isNaN(num) ? null : num;
  }

  private getInt(row: string[], field: string): number | null {
    if (!this.fieldIndices.has(field)) return null;
    const idx = this.fieldIndices.get(field)!;
    if (idx >= row.length) return null;
    const val = row[idx].trim();
    if (!val) return null;
    const num = Math.round(parseFloat(val));
    return isNaN(num) ? null : num;
  }

  private getBool(row: string[], field: string): boolean {
    if (!this.fieldIndices.has(field)) return false;
    const idx = this.fieldIndices.get(field)!;
    if (idx >= row.length) return false;
    const val = row[idx].trim().toLowerCase();
    return ['true', '1', 'yes', 'y', 'valid', 'ok'].includes(val);
  }

  private getDatetime(row: string[], field: string): string | null {
    if (!this.fieldIndices.has(field)) return null;
    const idx = this.fieldIndices.get(field)!;
    if (idx >= row.length) return null;
    return parseDatetime(row[idx]);
  }
}

export function parseCSVFile(content: string): TrackerRecord[] {
  const parser = new CSVParser();
  return parser.parseCSVContent(content);
}

File: logtail_dashboard/parser.py
"""CSV parser for tracker log files."""

import csv
import logging
from datetime import datetime
from io import StringIO
from typing import Optional

from .models import TrackerRecord

logger = logging.getLogger(__name__)

# Field name mappings for flexible CSV parsing
# Supports both dashboard format and GPS Tracker 2025 manual formats (Receiver CSV & SD Card CSV)
FIELD_MAPPINGS = {
    # Tracker ID: Report_StationID (receiver), User_LoggerID (SD card), or dashboard format
    "tracker_id": [
        "tracker_id", "id", "tracker", "device_id", "unit_id", "unique_id", "unique id",
        "report_stationid", "user_loggerid", "logger_id", "loggerid", "station_id", "stationid",
    ],
    # Time: Measurement_DateTime (manual) or dashboard format
    "time": [
        "time", "timestamp", "datetime", "time_local", "received_time",
        "time local received", "time_local_received",
        "measurement_datetime", "measurement_receiveddatetime",
    ],
    "time_gps": ["time_gps", "gps_time", "gps_timestamp"],
    # Position: GPS_lat, GPS_lon, GPS_alt (manual format)
    "lat": ["lat", "latitude", "gps_lat"],
    "lon": ["lon", "lng", "longitude", "gps_lon"],
    "alt": ["alt", "altitude", "alt_m", "altitude_m", "gps_alt"],
    # Speed/Course: GPS_SOG, GPS_COG (manual format - SOG in m/s, COG in degrees)
    "speed": ["speed", "speed_mps", "speed_m_s", "gps_sog", "sog"],
    "course": ["course", "heading", "course_deg", "gps_cog", "cog"],
    # GPS Quality: GPS_HDOP, GPS_FixValid, satellites (manual format)
    "hdop": ["hdop", "dop", "gps_hdop"],
    "satellites": ["satellites", "sats", "num_sats", "gps_satellites", "sat_count", "numsat"],
    "fix_valid": [
        "fix_valid", "fix", "gps_fix", "valid", "gps_fix_valid", "gps fix valid",
        "gps_fixvalid", "fixvalid",
    ],
    # Signal strength: RF_RSSI (manual format)
    "rssi": ["rssi", "rssi_dbm", "signal", "rf_rssi"],
    # Received time for latency calculation: Measurement_ReceivedDateTime (manual format)
    "time_received": [
        "time_received", "received_time", "rx_time", "receive_time",
        "measurement_receiveddatetime", "received_datetime",
    ],
    # Barometer: Barometer_Altitude, Barometer_Temperature, Barometer_Pressure (manual format)
    "baro_alt": [
        "baro_alt", "baro_altitude", "baro_alt_m", "barometric_altitude", "barometric altitude",
        "barometer_altitude",
    ],
    "baro_temp": ["baro_temp", "baro_temperature", "temp_c", "barometer_temperature"],
    "baro_press": ["baro_press", "baro_pressure", "pressure_hpa", "barometer_pressure"],
    # Battery voltage (SD card format): Battery_mV
    "battery_mv": ["battery_mv", "battery", "voltage_mv"],
}


class CSVParser:
    """Parser for tracker CSV log files."""

    def __init__(self):
        """Initialize parser."""
        self._field_indices: dict[str, int] = {}
        self._headers: list[str] = []

    def parse_csv(self, file_path) -> list[TrackerRecord]:
        """
        Parse a CSV file into tracker records.

        Args:
            file_path: Path to the CSV file.

        Returns:
            List of parsed and normalized tracker records.
        """
        try:
            from pathlib import Path
            path = Path(file_path)
            with open(path, "r", encoding="utf-8") as f:
                content = f.read()
                return self.parse_csv_content(content)
        except Exception as e:
            logger.error(f"Failed to read CSV file {file_path}: {e}")
            return []

    def parse_csv_content(self, content: str) -> list[TrackerRecord]:
        """
        Parse CSV content into tracker records.

        Args:
            content: CSV content as string.

        Returns:
            List of parsed and normalized tracker records.
        """
        records = []

        try:
            # Parse CSV
            reader = csv.reader(StringIO(content))
            rows = list(reader)

            if len(rows) < 2:
                # Need at least header + 1 data row
                return records

            # First row is header
            headers = [h.strip().lower() for h in rows[0]]
            self._headers = headers
            self._build_field_indices(headers)

            # Parse data rows
            for row_num, row in enumerate(rows[1:], start=2):
                try:
                    record = self._parse_row(row)
                    if record:
                        records.append(record)
                except Exception as e:
                    logger.warning(f"Failed to parse row {row_num}: {e}")
                    continue

        except Exception as e:
            logger.error(f"Failed to parse CSV content: {e}")

        return records

    def _build_field_indices(self, headers: list[str]) -> None:
        """
        Build index mapping from normalized field names to column indices.

        Args:
            headers: List of header names from CSV.
        """
        self._field_indices = {}

        for field_name, variants in FIELD_MAPPINGS.items():
            for variant in variants:
                if variant in headers:
                    self._field_indices[field_name] = headers.index(variant)
                    break

        logger.debug(f"Field indices: {self._field_indices}")

    def _parse_row(self, row: list[str]) -> Optional[TrackerRecord]:
        """
        Parse a single CSV row into a TrackerRecord.

        Args:
            row: List of values from CSV row.

        Returns:
            Normalized TrackerRecord or None if row is invalid.
        """
        # Must have tracker_id
        if "tracker_id" not in self._field_indices:
            logger.warning("CSV has no tracker_id column")
            return None

        tracker_id_idx = self._field_indices["tracker_id"]
        if tracker_id_idx >= len(row):
            return None

        tracker_id = row[tracker_id_idx].strip()
        if not tracker_id:
            return None

        # Parse time (fallback to now if not present)
        time_local = self._parse_datetime(row, "time")
        if time_local is None:
            time_local = datetime.now()

        # Parse GPS time
        time_gps = self._parse_datetime(row, "time_gps")

        # Parse received time (for latency calculation)
        time_received = self._parse_datetime(row, "time_received")

        # Parse position fields
        lat = self._parse_float(row, "lat")
        lon = self._parse_float(row, "lon")
        alt_m = self._parse_float(row, "alt")

        # Parse speed/course
        speed_mps = self._parse_float(row, "speed")
        course_deg = self._parse_float(row, "course")

        # Parse GPS quality
        hdop = self._parse_float(row, "hdop")
        satellites = self._parse_int(row, "satellites")
        fix_valid = self._parse_bool(row, "fix_valid")

        # Parse signal strength
        rssi_dbm = self._parse_float(row, "rssi")

        # Parse barometer data
        baro_alt_m = self._parse_float(row, "baro_alt")
        baro_temp_c = self._parse_float(row, "baro_temp")
        baro_press_hpa = self._parse_float(row, "baro_press")

        # Parse battery voltage (SD card format)
        battery_mv = self._parse_float(row, "battery_mv")

        # Calculate latency if both timestamps are available
        latency_ms = None
        if time_local and time_received:
            delta = time_received - time_local
            latency_ms = delta.total_seconds() * 1000

        return TrackerRecord(
            tracker_id=tracker_id,
            time_local_received=time_local,
            time_gps=time_gps,
            time_received=time_received,
            lat=lat,
            lon=lon,
            alt_m=alt_m,
            speed_mps=speed_mps,
            course_deg=course_deg,
            hdop=hdop,
            satellites=satellites,
            rssi_dbm=rssi_dbm,
            baro_alt_m=baro_alt_m,
            baro_temp_c=baro_temp_c,
            baro_press_hpa=baro_press_hpa,
            fix_valid=fix_valid,
            battery_mv=battery_mv,
            latency_ms=latency_ms,
        )

    def _parse_float(self, row: list[str], field_name: str) -> Optional[float]:
        """
        Parse a float value from row.

        Args:
            row: CSV row values.
            field_name: Normalized field name.

        Returns:
            Float value or None if not present/invalid.
        """
        if field_name not in self._field_indices:
            return None

        idx = self._field_indices[field_name]
        if idx >= len(row):
            return None

        value = row[idx].strip()
        if not value:
            return None

        try:
            return float(value)
        except ValueError:
            return None

    def _parse_int(self, row: list[str], field_name: str) -> Optional[int]:
        """
        Parse an integer value from row.

        Args:
            row: CSV row values.
            field_name: Normalized field name.

        Returns:
            Integer value or None if not present/invalid.
        """
        if field_name not in self._field_indices:
            return None

        idx = self._field_indices[field_name]
        if idx >= len(row):
            return None

        value = row[idx].strip()
        if not value:
            return None

        try:
            return int(float(value))  # Handle "7.0" -> 7
        except ValueError:
            return None

    def _parse_bool(self, row: list[str], field_name: str) -> bool:
        """
        Parse a boolean value from row.

        Args:
            row: CSV row values.
            field_name: Normalized field name.

        Returns:
            Boolean value (default False if not present/invalid).
        """
        if field_name not in self._field_indices:
            return False

        idx = self._field_indices[field_name]
        if idx >= len(row):
            return False

        value = row[idx].strip().lower()
        if not value:
            return False

        # Handle various boolean representations
        return value in ["true", "1", "yes", "y", "valid", "ok"]

    def _parse_datetime(self, row: list[str], field_name: str) -> Optional[datetime]:
        """
        Parse a datetime value from row.

        Args:
            row: CSV row values.
            field_name: Normalized field name.

        Returns:
            Datetime value or None if not present/invalid.
        """
        if field_name not in self._field_indices:
            return None

        idx = self._field_indices[field_name]
        if idx >= len(row):
            return None

        value = row[idx].strip()
        if not value:
            return None

        # Try common datetime formats
        formats = [
            "%Y-%m-%dT%H:%M:%S",
            "%Y-%m-%d %H:%M:%S",
            "%Y-%m-%dT%H:%M:%S.%f",
            "%Y-%m-%d %H:%M:%S.%f",
            "%Y-%m-%dT%H:%M:%SZ",
            "%m/%d/%Y %H:%M:%S",
            "%d/%m/%Y %H:%M:%S",
        ]

        for fmt in formats:
            try:
                return datetime.strptime(value, fmt)
            except ValueError:
                continue

        # Try ISO format with timezone
        try:
            # Remove timezone info for simplicity
            if "+" in value or value.endswith("Z"):
                value = value.split("+")[0].split("Z")[0]
                return datetime.fromisoformat(value)
        except ValueError:
            pass

        logger.warning(f"Could not parse datetime: {value}")
        return None


def parse_csv_file(file_path: str) -> list[TrackerRecord]:
    """
    Parse a complete CSV file.

    Args:
        file_path: Path to CSV file.

    Returns:
        List of parsed tracker records.
    """
    parser = CSVParser()

    try:
        with open(file_path, "r", encoding="utf-8") as f:
            content = f.read()
            return parser.parse_csv_content(content)
    except Exception as e:
        logger.error(f"Failed to read CSV file {file_path}: {e}")
        return []


# Alias for backward compatibility
Parser = CSVParser
